app/scripts/controllers/deployment.js

```diff
@@ -131,28 +156,23 @@ angular.module('openshiftConsole')
                   message: "This deployment has been deleted."
                 };
               }
+
+              var previous = $scope.deployment;
               $scope.deployment = deployment;
               $scope.updatingPausedState = false;
+              updateHPAWarnings();
 
-              if ($scope.forms.deploymentEnvVars.$pristine) {
-                copyDeploymentAndEnsureEnv(deployment);
+              updateEnvironment(deployment, previous);
+
+              // Wait for a pending save to complete to avoid a race between the PUT and the watch callbacks.
+              if (saveEnvPromise) {
+                saveEnvPromise.finally(function() {
+                  updateEnvironment(deployment, previous);
```

Comment:
@jwforres It looks like calling `updateEnvironment` from `saveEnvPromise.finally` is enough.

* If the PUT succeeds, the success callback in `saveEnvVars` will set the form to pristine before `updateEnvironment` is called. When the form is pristine, `updateEnvironment` will simply copy the new values.
* If the PUT fails, the normal conflict checking is performed by `updateEnvironment`.